File: frontend/src/lib/api/teams.ts
import type { WorkspaceMemberRow } from "./workspace";
import { apiFetch } from "./client";

export interface TeamMemberPreview {
  id: string;
  fullName: string;
  avatarUrl?: string | null;
  role: string;
}

export interface TeamSummary {
  id: string;
  name: string;
  color: string;
  icon: string;
  description?: string | null;
  memberCount: number;
  membersPreview: TeamMemberPreview[];
  createdAt: string | null;
  createdBy: { id: string; fullName: string; avatarUrl?: string | null } | null;
}

export interface TeamDetail extends TeamSummary {
  members: TeamMemberPreview[];
  updatedAt: string | null;
}

export interface WorkspaceMemberWithTeams extends WorkspaceMemberRow {
  teams?: { id: string; name: string; color: string; icon: string }[];
}

function wsPath(workspaceId: string, path: string) {
  return `/workspaces/${workspaceId}${path}`;
}

export function fetchTeams(
  token: string,
  workspaceId: string,
  params?: { sort?: string; order?: string }
) {
  const qs = new URLSearchParams();
  if (params?.sort) qs.set("sort", params.sort);
  if (params?.order) qs.set("order", params.order);
  const query = qs.toString();
  return apiFetch<{ data: TeamSummary[] }>(
    wsPath(workspaceId, `/teams${query ? `?${query}` : ""}`),
    { token }
  );
}

export function fetchMyTeams(token: string, workspaceId: string) {
  return apiFetch<{ data: TeamSummary[] }>(
    wsPath(workspaceId, "/teams/mine"),
    { token }
  );
}

export function fetchTeam(token: string, workspaceId: string, teamId: string) {
  return apiFetch<TeamDetail>(wsPath(workspaceId, `/teams/${teamId}`), {
    token,
  });
}

export function createTeam(
  token: string,
  workspaceId: string,
  body: {
    name: string;
    color?: string;
    icon?: string;
    description?: string;
    memberIds?: string[];
  }
) {
  return apiFetch<TeamDetail>(wsPath(workspaceId, "/teams"), {
    method: "POST",
    token,
    body: JSON.stringify(body),
  });
}

export function updateTeam(
  token: string,
  workspaceId: string,
  teamId: string,
  body: {
    name?: string;
    color?: string;
    icon?: string;
    description?: string;
  }
) {
  return apiFetch<TeamDetail>(wsPath(workspaceId, `/teams/${teamId}`), {
    method: "PATCH",
    token,
    body: JSON.stringify(body),
  });
}

export function deleteTeam(
  token: string,
  workspaceId: string,
  teamId: string
) {
  return apiFetch<{ ok: boolean }>(
    wsPath(workspaceId, `/teams/${teamId}`),
    { method: "DELETE", token }
  );
}

export function addTeamMember(
  token: string,
  workspaceId: string,
  teamId: string,
  userId: string,
  role: "LEAD" | "MEMBER" = "MEMBER"
) {
  return apiFetch<TeamDetail>(
    wsPath(workspaceId, `/teams/${teamId}/members`),
    {
      method: "POST",
      token,
      body: JSON.stringify({ userId, role }),
    }
  );
}

export function removeTeamMember(
  token: string,
  workspaceId: string,
  teamId: string,
  userId: string
) {
  return apiFetch<TeamDetail>(
    wsPath(workspaceId, `/teams/${teamId}/members/${userId}`),
    { method: "DELETE", token }
  );
}

File: backend-py/app/api/v1/teams.py
from fastapi import APIRouter, Query, status

from app.deps.auth import CurrentUserDep, DbSession
from app.deps.workspace import WorkspaceMemberDep
from app.schemas.team import AddTeamMemberBody, CreateTeamBody, UpdateTeamBody
from app.services import team_service

router = APIRouter(prefix="/workspaces/{workspace_id}/teams", tags=["teams"])


@router.get("")
async def list_teams(
    workspace_id: str,
    session: DbSession,
    user: CurrentUserDep,
    _member: WorkspaceMemberDep,
    sort: str = Query("name", pattern="^(name|created|members)$"),
    order: str = Query("asc", pattern="^(asc|desc)$"),
):
    data = await team_service.list_teams(
        session, workspace_id, sort=sort, order=order
    )
    return {"data": data}


@router.get("/mine")
async def list_my_teams(
    workspace_id: str,
    session: DbSession,
    user: CurrentUserDep,
    _member: WorkspaceMemberDep,
):
    data = await team_service.list_my_teams(session, workspace_id, user.id)
    return {"data": data}


@router.post("", status_code=status.HTTP_201_CREATED)
async def create_team(
    body: CreateTeamBody,
    workspace_id: str,
    session: DbSession,
    user: CurrentUserDep,
    _member: WorkspaceMemberDep,
):
    return await team_service.create_team(session, workspace_id, user.id, body)


@router.get("/{team_id}")
async def get_team(
    workspace_id: str,
    team_id: str,
    session: DbSession,
    user: CurrentUserDep,
    _member: WorkspaceMemberDep,
):
    return await team_service.get_team(session, workspace_id, team_id)


@router.patch("/{team_id}")
async def update_team(
    body: UpdateTeamBody,
    workspace_id: str,
    team_id: str,
    session: DbSession,
    user: CurrentUserDep,
    member: WorkspaceMemberDep,
):
    return await team_service.update_team(
        session, workspace_id, team_id, user.id, member.role, body
    )


@router.delete("/{team_id}")
async def delete_team(
    workspace_id: str,
    team_id: str,
    session: DbSession,
    user: CurrentUserDep,
    member: WorkspaceMemberDep,
):
    return await team_service.delete_team(
        session, workspace_id, team_id, user.id, member.role
    )


@router.post("/{team_id}/members")
async def add_team_member(
    body: AddTeamMemberBody,
    workspace_id: str,
    team_id: str,
    session: DbSession,
    user: CurrentUserDep,
    member: WorkspaceMemberDep,
):
    return await team_service.add_team_member(
        session, workspace_id, team_id, user.id, member.role, body
    )


@router.delete("/{team_id}/members/{target_user_id}")
async def remove_team_member(
    workspace_id: str,
    team_id: str,
    target_user_id: str,
    session: DbSession,
    user: CurrentUserDep,
    member: WorkspaceMemberDep,
):
    return await team_service.remove_team_member(
        session, workspace_id, team_id, target_user_id, user.id, member.role
    )
